File: StriverSDE/DynamicProgramming/ms_1.cpp
// you can use includes, for example:
// #include <algorithm>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;
vector<vector<int>> dp;

int recur(vector<int>& h,int pos,int r_x, int r_y,int n){    //r_x: remaining time in x r_y: remaining time 
    if(pos>=n) return 0;
    if(dp[pos][abs(r_x-r_y)]!=-1) return dp[pos][abs(r_x-r_y)];
    int v1=0,v2=0,v3=0;
    if(r_x>=h[pos]){
        v1=1+recur(h,pos+1,r_x-h[pos],r_y,n);
    }
    if(r_y>=h[pos]){
        v2=1+recur(h,pos+1,r_x,r_y-h[pos],n);
    }
    v3=recur(h,pos+1,r_x,r_y,n);
    return dp[pos][abs(r_x-r_y)]=max(v1,max(v2,v3));
}

int solution(vector<int> &H, int X, int Y) {
    int N = H.size();
    dp.clear();
    dp.resize(N, vector<int>(2 * (X + Y) + 1, -1));
    return recur(H, 0, X, Y, N); 
}

File: StriverSDE/Recursion/ms_1.java
// you can also use imports, for example:
import java.util.*;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");

class Solution {
    public int solution(int[] A) {
        A.sort() ;
        Set<Integer> s = new HashSet<>() ;

    for(int i = 0 ; i < A.length ; i++ ){
        for(int j = i +1 ; j < A.length ; j++ ){
            int diff = A[i] - A[j] ;
            set.add(diff) ;
        }
    }
    int ans = Integer.MIN_VALUE ;
    int[] arr = set.toArray() ; 
    for(int i : arr) {
        ans=Math.max(ans,fun(i,A)); 
    }
    return ans ;
    }

     public static int fun(int d, int[] A} {
        HashMap<Integer,Integer> mp = new HashMap<>() ;
        int ans= Integer.MIN_VALUE ;
        for(int i=0;i<n;i++){
            int x=A[i]-d;
            if(mp.get(x)>0){
                mp.put(A[i], mp.get(x) +1) ;
            }
            else{
                mp.put(A[i],1);
            }
            ans=Math.max(ans,mp[A[i]]);
        }
        return ans;
}

